src/software/ai/ai.cpp

```diff
@@ -20,39 +21,58 @@ void AI::overridePlay(std::unique_ptr<Play> play)
     override_play = std::move(play);
 }
 
-void AI::overridePlayFromName(std::string name)
+void AI::overridePlayFromProto(TbotsProto::Play play_proto)
 {
-    overridePlay(GenericFactory<std::string, Play, AiConfig>::create(name, ai_config));
+    overridePlay(std::move(createPlay(play_proto, ai_config_)));
+}
+
+void AI::updateAiConfig(TbotsProto::AiConfig ai_config)
+{
+    ai_config_        = ai_config;
+    ai_config_changed = true;
 }
 
 void AI::checkAiConfig()
 {
-    bool current_override = ai_config->getAiControlConfig()->getOverrideAiPlay()->value();
-    std::string current_override_name =
-        ai_config->getAiControlConfig()->getCurrentAiPlay()->value();
-    if (current_override && (current_override != prev_override ||
-                             current_override_name != prev_override_name))
+    auto current_override = ai_config_.ai_control_config().override_ai_play();
+
+    // If we have a new override, and its not back to the Ai selection,
+    // lets override the play
+    if (current_override != prev_override &&
+        current_override != TbotsProto::PlayName::UseAiSelection)
+    {
+        TbotsProto::Play play_proto;
+        play_proto.set_name(current_override);
+        overridePlayFromProto(play_proto);
+    }
+
+    // If we have a new override but its back to the Ai selection, lets
+    // clear the override
+    if (current_override != prev_override &&
+        current_override == TbotsProto::PlayName::UseAiSelection)
     {
-        overridePlayFromName(
-            ai_config->getAiControlConfig()->getCurrentAiPlay()->value());
+        overridePlay(nullptr);
     }
-    prev_override      = current_override;
-    prev_override_name = current_override_name;
+    prev_override = current_override;
 }
 
 std::unique_ptr<TbotsProto::PrimitiveSet> AI::getPrimitives(const World& world)
 {
     checkAiConfig();
     fsm->process_event(PlaySelectionFSM::Update(
         [this](std::unique_ptr<Play> play) { current_play = std::move(play); },
-        world.gameState()));
+        world.gameState(), ai_config_));
 
-    if (!field_to_path_planner_factory.contains(world.field()))
+    // We construct the global path planner once for the field. If the AI Config
+    // changes, there might be an update we need to reconstruct the path planner
+    // for and propagate the parameter change.
+    if (!field_to_path_planner_factory.contains(world.field()) || ai_config_changed)
     {
-        field_to_path_planner_factory.emplace(
+        field_to_path_planner_factory.insert_or_assign(
             world.field(),
-            GlobalPathPlannerFactory(ai_config->getRobotNavigationObstacleConfig(),
+            GlobalPathPlannerFactory(ai_config_.robot_navigation_obstacle_config(),
```

Comment:
Is the `GlobalPathPlannerFactory` the only class that needs to be updated with the new config? Do we not need to update any of the plays (i.e. the current play)? Or is the idea that once we switch to a new play and it's constructed, then it will get the updated config values?

Reply:
so the main modules we have that need to be updated are:

sensor_fusion
play_selection_fsm (which will handle tactics, plays, pass gen, etc..)
path_planner

sensor fusion will reset itself if the sensor fusion config changes
path planner reconstructs itself when the ai_config changes

and finally the playselection fsm will receive an update event with the ai_config

that being said, when I went to go take a closer look, the way I had it setup rn the play selection fsm would have only updated the ai_config when the play changed

so I've added a line of code to restart the playselection fsm if the ai config changes